common/src/main/java/io/pravega/common/util/ResourcePool.java

```diff
@@ -281,6 +279,48 @@ public void shutdown() {
 
     @Data
     private static class WaitingRequest<T> {
-        private final CompletableFuture<T> future;
+        private final CompletableFuture<ClosableResource<T>> future;
+    }
+
+    /**
+     * A closeable resource wrapper class which returns the resource back to the pool automatically once it is closed.
+     * Its close method is idempotent and can be invoked multiple times without returning the resource more than once. 
+     * @param <T> Type of underlying resource
+     */
+    public static class ClosableResource<T> implements AutoCloseable {
+        private final ResourcePool<T> resourcePool;
+        private final T resource;
+        private final AtomicBoolean invalid;
+        @GuardedBy("$lock")
+        private boolean isClosed;
+        private ClosableResource(T resource, ResourcePool<T> resourcePool) {
+            this.resourcePool = resourcePool;
+            this.resource = resource;
+            this.invalid = new AtomicBoolean(false);
+            this.isClosed = false;
+        }
+
+        public T getResource() {
+            return resource;
+        }
+        
+        public void invalidate() {
+            invalid.set(true);
+        }
+        
+        @Override
+        @Synchronized
+        public void close() {
+            // Close is idempotent. 
+            // If close had already been invoked on this resource wrapper, then we do not return the resource to the pool.  
+            if (!isClosed) {
+                if (invalid.get()) {
+                    resourcePool.returnResource(resource, false);
```

Comment:
1. `resourcePool.returnResource(resource, !invalid.get())`. That way you can also eliminate `returnResource()` (with no args).

2. Be very careful with this `synchronized` method. You're acquiring a lock here and acquiring a different lock in `resourcePool.returnResource`. If not used carefully, this could lead to a deadlock. You are completing a future in `returnResource` while holding this sub-class' lock, so that is a a potential cause for concern. Would we be able to solve the same problem by changing `isClosed` to an `AtomicBoolean` and using `.compareAndSet(false, true)` to ensure `close` is called only once?

Reply:
done